libs/common-types/src/lib.rs

```diff
@@ -0,0 +1,270 @@
+			PoolRole::RiskAdmin => self.admin.contains(AdminRoles::RISK_ADMIN),
+			PoolRole::TrancheInvestor(id, _) => self.tranche_investor.contains(id),
+		}
+	}
+
+	fn empty(&self) -> bool {
+		self.admin.is_empty() && self.tranche_investor.is_empty()
+	}
+
+	fn rm(&mut self, property: Self::Property) {
+		match property {
+			PoolRole::Borrower => self.admin.remove(AdminRoles::BORROWER),
+			PoolRole::LiquidityAdmin => self.admin.remove(AdminRoles::LIQUIDITY_ADMIN),
+			PoolRole::PoolAdmin => self.admin.remove(AdminRoles::POOL_ADMIN),
+			PoolRole::PricingAdmin => self.admin.remove(AdminRoles::PRICING_ADMIN),
+			PoolRole::MemberListAdmin => self.admin.remove(AdminRoles::MEMBER_LIST_ADMIN),
+			PoolRole::RiskAdmin => self.admin.remove(AdminRoles::RISK_ADMIN),
+			PoolRole::TrancheInvestor(id, delta) => self.tranche_investor.remove(id, delta),
+		}
+	}
+
+	fn add(&mut self, property: Self::Property) {
+		match property {
+			PoolRole::Borrower => self.admin.insert(AdminRoles::BORROWER),
+			PoolRole::LiquidityAdmin => self.admin.insert(AdminRoles::LIQUIDITY_ADMIN),
+			PoolRole::PoolAdmin => self.admin.insert(AdminRoles::POOL_ADMIN),
+			PoolRole::PricingAdmin => self.admin.insert(AdminRoles::PRICING_ADMIN),
+			PoolRole::MemberListAdmin => self.admin.insert(AdminRoles::MEMBER_LIST_ADMIN),
+			PoolRole::RiskAdmin => self.admin.insert(AdminRoles::RISK_ADMIN),
+			PoolRole::TrancheInvestor(id, delta) => self.tranche_investor.insert(id, delta),
+		}
+	}
+}
+
+impl<Now, MaxTranches, MinDelay, TrancheId, Moment>
+	TrancheInvestors<Now, MaxTranches, MinDelay, TrancheId, Moment>
+where
+	MaxTranches: Get<TrancheId>,
+	Now: Time<Moment = Moment>,
+	MinDelay: Get<Moment>,
+	Moment: PartialEq + PartialOrd + Saturating + Ord + Copy,
+	TrancheId: PartialEq + PartialOrd,
+{
+	pub fn empty() -> Self {
+		Self::default()
+	}
+
+	pub fn is_empty(&self) -> bool {
+		self.info.is_empty()
+	}
+
+	fn validity(&self, delta: Moment) -> Moment {
+		let now = Now::now();
+		let min_validity = now.saturating_add(MinDelay::get());
+		let req_validity = now.saturating_add(delta);
+		max(min_validity, req_validity)
+	}
+
+	pub fn contains(&self, tranche: TrancheId) -> bool {
+		if tranche >= self.max_tranches {
+			return false;
+		}
+
+		self.info
+			.iter()
+			.position(|info| info.tranche_id == tranche && info.permissioned_till >= Now::now())
+			.is_some()
+	}
+
+	pub fn remove(&mut self, tranche: TrancheId, delta: Moment) {
+		if tranche >= self.max_tranches {
+			return;
+		}
+
+		if let Some(index) = self.info.iter().position(|info| info.tranche_id == tranche) {
+			let valid_till = &self.info[index].permissioned_till;
+			let now = Now::now();
+
+			if *valid_till <= now {
+				// The account is already invalid. Hence no more grace period
+				return;
+			} else {
+				// Ensure that permissioned_till is at least now + min_delay.
+				self.info[index].permissioned_till = self.validity(delta);
```

Comment:
Why is this argument called `delta`? If I'm not mistaken, it's the new "permissioned until timestamp". `delta` implies that it's the change in the value, as in, I would understand that to be how much we're adding/removing to the current value (i.e. if you pass the number of seconds in a day, that it moves up by one day).

Reply:
Also, if someone attempts to set it to less than `now + min_delay,` could we throw an error here instead of setting it to exactly `now + min_delay`? I think that's better to block than to implicitly have different behaviour.

Reply:
It is currently as you are understanding it. It moves the permission up by max(now + delta , now + min_delay). I don't have an opinion here. I can change it to be `permissioned_until_timestamp`. Should this be the default for every update (i.e. insert also). 

Erroring out upon not respecting the `MinDelay` is a bigger change. I understand your reasoning. The implicit behavior is not good. 

I will add this.

Reply:
Hm, on second thought, I think you're right that using `delta` here makes a lot of sense. The user would indeed whitelist someone "for a year", not "until day x".  Let's keep it with the delta variable 👍 

Changing it to error out would be great.